clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/GoogleUserDataProvider.groovy

```diff
@@ -0,0 +1,139 @@
+                  BasicGoogleDeployDescription description,
+                  GoogleNamedAccountCredentials credentials, String customUserData) {
+    String userDataFile = credentials.getUserDataFile()
+    String rawUserData = getFileContents(userDataFile)
+    String commonUserData = replaceTokens(rawUserData, description, serverGroupName, instanceTemplateName)
+
+    def userDataMap = stringToUserDataMap(commonUserData)
+
+    if (customUserData){
+      def customUserDataMap = stringToUserDataMap(customUserData)
+      userDataMap << customUserDataMap
+    }
+    userDataMap
+  }
+
+  /**
+   * Returns the contents of a file or an empty string if the file doesn't exist.
+   */
+  @PackageScope
+  String getFileContents(String filename) {
+    if (!filename) {
+      return ''
+    }
+    try {
+      File file = new File(filename)
+      def rawContentsList = file.readLines()
+      def contentsList = []
+      for (line in rawContentsList) {
+        if (!line.startsWith('#')) {
+          contentsList = contentsList << line
+        }
+      }
+      String contents = contentsList.join(',')
+      return contents
+    } catch (IOException e) {
+      log.warn("Failed to read user data file ${filename}; ${e.message}")
+      return ''
+    }
+  }
+  /**
+   * Returns the user data with the tokens replaced.
+   *
+   * Currently supports the following tokens:
+   *
+   * %%account%% 	    the name of the account
+   * %%accounttype%% 	the accountType of the account
+   * %%env%%        	the environment of the account
+   * %%app%%          the name of the app
+   * %%region%% 	    the deployment region
+   * %%group%% 	      the name of the server group
+   * %%autogrp%% 	    the name of the server group
+   * %%cluster%% 	    the name of the cluster
+   * %%stack%% 	      the stack component of the cluster name
+   * %%detail%% 	    the detail component of the cluster name
+   * %%launchconfig%% the name of the instance template
+   */
+  private String replaceTokens(String rawUserData, BasicGoogleDeployDescription description,
+                               String serverGroupName, String instanceTemplateName) {
+    if (!rawUserData) {
+      return ''
+    }
+    Names names = Names.parseName(serverGroupName)
+    // Replace the tokens & return the result
+    String result = rawUserData
+      .replace('%%account%%', description.accountName)
+      .replace('%%accounttype%%', description.credentials.accountType ?: '')
+      .replace('%%app%%', names.app ?: '')
+      .replace('%%env%%', description.credentials.environment ?: '')
+      .replace('%%region%%', description.region ?: '')
+      .replace('%%stack%%', description.stack ?: '')
+      .replace('%%group%%', names.group ?: '')
+      .replace('%%autogrp%%', names.group ?: '')
+      .replace('%%cluster%%', names.cluster ?: '')
+      .replace('%%detail%%', names.detail ?: '')
+      .replace('%%launchconfig%%', instanceTemplateName ?: '')
+
+    result
```

Comment:
While groovy lets you do this, we generally prefer to add `return` statements to functions > 1 line long